### ISSUE TITLE
Consider disallowing forward references to non-global names in `__future__.annotations` (for compatibility)

### ISSUE THREAD

**Issue by BHSPitMonkey** [astral-sh/ty#1747]
### Summary

Examples: https://play.ty.dev/e6628610-7be6-4090-82ca-d1cadf92e85f

Beginning with the `alpha.28` release (and still happening as of `alpha.30`), ty emits `invalid-type-form` for valid types that happen to share a name with an instance method.

Example involving the `set` builtin:

```python
class Foo:
  def set(self) -> None:
    pass

  def bar(self) -> set[int]:  # invalid-type-form
    return {1, 2, 3}

  def baz(self) -> set:       # invalid-type-form
    return {"hello", "world"}
```

```
Invalid subscript of object of type `def set(self) -> None` in type expression
Variable of type `def set(self) -> None` is not allowed in a type expression
```

Example using a non-builtin type:

```python
from datetime import datetime

class Foo:
  def datetime(self) -> None:
    pass

  def maybe_return_date(self) -> datetime | None:  # invalid-type-form
    return None
```

```
Variable of type `def datetime(self) -> None` is not allowed
```

### Version

ty 0.0.1-alpha.28, ty 0.0.1-alpha.29, ty 0.0.1-alpha.30

**Comment by sharkdp** [comment]
Thank you for reporting this.

Pyright seems to support this, but mypy and pyrefly also think that something is wrong here. I think it's reasonable for `ty` to assume that `set` refers to the *symbol in the same scope of the same name*, even if that happens to be a function definition. I would advise to disambiguate the name of the type by using an `as`-import, a type alias, a fully qualified name or similar.

**Comment by carljm** [comment]
Yes, this is an intentional choice in ty to respect normal Python name-resolution rules for names in type annotations, rather than special-casing some logic to skip the local scope and prefer the global scope. The ty behavior also matches the runtime behavior of resolving `__annotations__`:

```
>>> class Foo:
...   def set(self) -> None:
...     pass
...
...   def bar(self) -> set[int]:  # invalid-type-form
...     return {1, 2, 3}
...
Traceback (most recent call last):
  File "<python-input-0>", line 1, in <module>
    class Foo:
    ...<4 lines>...
        return {1, 2, 3}
  File "<python-input-0>", line 5, in Foo
    def bar(self) -> set[int]:  # invalid-type-form
                     ~~~^^^^^
TypeError: 'function' object is not subscriptable
```

IMO it's unfortunate that there has been confusion in the past around these name resolution rules (contributed to both by type-checkers and the behavior of `typing.get_type_hints` with string annotations), but I think it's clear that using normal resolution rules is the least confusing and most consistent path forward.

**Comment by BHSPitMonkey** [comment]
Thanks for clarifying, that makes sense! I only thought to compare against Pyright's behavior, but I see what you mean about adhering to how Python treats the (class) local scope here. I'll get around this by using `-> builtins.set[...]` in our code.

**Comment by correctmost** [comment]
I have seen a similar `invalid-type-form` warning in the archinstall codebase where ty is the only type checker to issue a warning:

```python
from __future__ import annotations
import uuid
from enum import Enum


class PartitionGUID(Enum):
    @property
    def bytes(self) -> bytes:  # error[invalid-type-form]: Variable of type `property` is not allowed in a type expression
        return uuid.UUID(self.value).bytes
```

Pyrefly, Pyright, mypy, and zuban do not emit type-form warnings for the same code, and the code also works at runtime.  (ty does not warn about the code if `from __future__ import annotations` is removed.)

I don't know if that possibly changes this bug's resolution, but I thought I'd at least mention a case where ty seems to differ in behavior from other major type checkers.

**Comment by BHSPitMonkey** [comment]
@correctmost That's the same case discussed above; It works at runtime because Python doesn't enforce type annotations, but in the `class PartitionGUID` scope the Python runtime would interpret `bytes` as a reference to the property instead of the builtin (e.g. if you added a class variable `foo = bytes`, it would refer to the property).

You can use the same workaround I'm using, e.g.:

```python
from __future__ import annotations
import builtins
import uuid
from enum import Enum


class PartitionGUID(Enum):
    @property
    def bytes(self) -> builtins.bytes:  # No error
        return uuid.UUID(self.value).bytes
```

**Comment by carljm** [comment]
Thanks @correctmost for the additional example!

For stringified annotations (`from __future__ import annotations`), runtime never even evaluates the annotation. `typing.get_type_hints` will evaluate the stringified annotation, but it has a special-case internally to swap the local and global namespaces, so the global namespace is preferred over the local one.

This example works for mypy, pyrefly, and zuban for an entirely different reason than it "works" in `get_type_hints`. They do not implement the same namespace-swapping that `get_type_hints` does (and pyright does), which is clear because they all complain about the annotation on `foo` here:

```py
from __future__ import annotations

class C:
    def bytes(self) -> bytes:  # mypy, pyrefly, zuban are all OK with this
        return b""

    def foo(self) -> bytes:  # mypy, pyrefly, zuban all error here
        return self.bytes()
```

At runtime, names in type annotations with `from __future__ import annotations` are only evaluated if you explicitly ask for their evaluation, and not immediately. This means they can be forward references. Ty models this accurately, and thus (with `from __future__ import annotations`) we consider the annotation `bytes` in `def bytes(self) -> bytes:` to be a forward-reference to that same method.

Note that in Python 3.14, with PEP 649/749, annotations are deferred by default (without stringifying them), and `from __future__ import annotations` will become deprecated in future. And the behavior of Python 3.14 at runtime matches ty's behavior:

```
Python 3.14.0rc2 (main, Aug 28 2025, 17:02:21) [Clang 20.1.4 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> class C:
...     def bytes(self) -> bytes:
...         return b""
...
>>> C.bytes.__annotations__
{'return': <function C.bytes at 0x1048607d0>}
```

So it still seems to me that ty's behavior here is the most consistent, and that the best thing to do with such annotations is to disambiguate them, which will give full compatibility across all type checkers and future runtime behavior.

**Comment by carljm** [comment]
Since this is clearly a subtle issue that other people are likely to run into, and there's at least some possibility that we might decide to make our behavior less internally consistent in order to improve compatibility with other type checkers (though at the moment I'm not convinced we should do this), I'll go ahead and reopen this for visibility.

**Comment by carljm** [comment]
To better illustrate the difference between ty and other type checkers, ty supports this forward reference (in Python 3.14, or if `from __future__ import annotations` is active):

```py
class C:
    def foo(self) -> IntAlias:
        return 1

    IntAlias = int
```

Whereas mypy/pyrefly/zuban all error that `IntAlias` is not defined (but are fine if `IntAlias` is defined above `foo` instead). This is what leads to the difference in behavior on @correctmost 's example: ty is the only one to support forward references in non-global scope.

But at runtime in Python 3.14, the above works fine and the return annotation in `__annotations__` resolves to `int`.

**Comment by MichaReiser** [comment]
It might be worth improving our diagnostics if we dedect this very specific issue

**Comment by correctmost** [comment]
> This example works for mypy, pyrefly, and zuban for an entirely different reason than it "works" in `get_type_hints`. They do not implement the same namespace-swapping that `get_type_hints` does (and pyright does)

Thank you for the detailed explanation about the various subtleties :).

**Comment by konn** [comment]
Hi, I'm just redirected from #2101. We have a
case where all the classes/functions are defined in the same module, and that stub file has thousands of auto-generated lines.

Much simplified version is as follows:

```py
class A: ...

class B:
    def A(self, name: str) -> A: ...
    def take_a(self, a: A) -> str: ...
    def return_a(self) -> A: ...
```

In this monolithic module case, I think there is no way to refer class `A` defined toplevel inside class B, IIUC. Since these codes are machine-generated, it is rather hard to reorder declarations. I think it does still make sense to provide some workaround for this case.

**Comment by carljm** [comment]
@konn In your original example, every type checker except pyright will error on the return annotation of `return_a` method, see e.g. mypy: https://mypy-play.net/?mypy=latest&python=3.12&gist=032b3775e8d655314c535884b76b36e2

Pyright doesn't error because it copies the local/global namespace swap performed by `typing.get_type_hints`. It does this even when `from __future__ import annotations` is not active, which means it gives results which don't match the runtime resolution of annotations.

The workaround available in your case (or any similar case) is to alias the name to remove ambiguity:

```py
class A: ...
_AliasToA = A

class B:
    def A(self, name: str) -> _AliasToA: ...
    def take_a(self, a: A) -> str: ...
    def return_a(self) -> _AliasToA: ...
```

**Comment by nils-borrmann-tacto** [comment]
I want to add that ty is also inconsistent with ruff's rule A003: https://github.com/astral-sh/ruff/issues/7806

**Comment by carljm** [comment]
@nils-borrmann-tacto Not sure I understand where there is inconsistency. A003 complains if you have a class attribute which shadows a builtin name, and you reference the shadowing name in an annotation. These are the same scenarios where ty resolves the annotation to the shadowing name (as is the runtime behavior in 3.14+). This seems consistent to me -- both ruff and ty are discouraging this shadowing, and requiring disambiguation. Can you clarify where you see an inconsistency?

**Comment by nils-borrmann-tacto** [comment]
@carljm Using a modified example from the ruff docs:

Ty emits an error here. Ruff allows this. Ruff would only fail, when using `list` further down in the class (this behaviour was changed in the ticket I linked above).

```py
class Class:
    @staticmethod
    def list() -> list[None]:
        return []
```

**Comment by carljm** [comment]
@nils-borrmann-tacto I see, thanks for the example. The difference is that Ruff does not account for the Python 3.14+ behavior where evaluation of annotations is deferred. If you check this example in ty set to Python 3.13, it won't error; but if you select Python 3.14, it will error. This is accurate to the runtime behavior in 3.13 and 3.14:

```pycon
➜ uv run -p python3.13 python
Python 3.13.2 (main, Mar 11 2025, 17:30:09) [Clang 20.1.0 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> class Class:
...     @staticmethod
...     def list() -> list[None]:
...         return []
...
>>> Class.list.__annotations__
{'return': list[None]}

➜ uv run -p python3.14 python
Python 3.14.3 (main, Feb 12 2026, 00:29:58) [Clang 21.1.4 ] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> class Class:
...     @staticmethod
...     def list() -> list[None]:
...         return []
...
>>> Class.list.__annotations__
Traceback (most recent call last):
  File "<python-input-1>", line 1, in <module>
    Class.list.__annotations__
  File "<python-input-0>", line 3, in __annotate__
    def list() -> list[None]:
                  ~~~~^^^^^^
TypeError: 'staticmethod' object is not subscriptable
```

I think it would be consistent to the spirit of the Ruff implementation of A003 for it to add support for erroring on this case, under Python 3.14+.

**Comment by anuraaga** [comment]
Thanks a lot for the discussion, I also ran into this issue in the same form as https://github.com/astral-sh/ty/issues/1747#issuecomment-4096894705, just a simple built-in return type of a method with the same name (for me, `def type(self) -> type[Message]`). I use `__future__` annotations - actually I was surprised the behavior is different on 3.14 without the future import so just learned that. 

To understand what to expect when switching over from pyright, I'm a bit unclear with what the general direction is for this issue - will it continue to be an error even with `__future__` annotations, which doesn't match the runtime behavior? Without the import, it sounds like it will continue to be an error to match 3.14 behavior which seems good to me, if a downstream user has a chance of seeing an error, then it should be an error regardless of the current running interpreter. But it seems runtime behavior always aligns with no error when the future import is there.

**Comment by carljm** [comment]
@anuraaga It's a bit hard to say what the "runtime behavior" of `__future__.annotations` is, since by default it doesn't do name resolution at all, annotations are just strings. If you mean the runtime behavior of `typing.get_type_hints`, which privileges the global scope over the local scope in a way that Python runtime never does for name lookup, this issue suggests that we could consider doing something similar to that for `__future__.annotations`. We haven't made the decision to do that, though. It's not very attractive to put extra work into modeling several different semantics, when PEP 649/749 semantics are the future, and `__future__.annotations` is deprecated.

**Comment by knutwannheden** [comment]
A data point for the decision: this makes part of pydantic's public API resolve to `Unknown` for every pydantic user, with no diagnostic they can see.

`pydantic/main.py` uses `from __future__ import annotations`, and `BaseModel` defines a (deprecated) method named `dict`. So the return annotation of `BaseModel.model_fields` and `model_computed_fields`, `dict[str, FieldInfo]`, resolves to that method:

```py
from pydantic import BaseModel

class A(BaseModel):
    a: str

reveal_type(A.model_fields)           # ty 0.0.83: Unknown
reveal_type(A.model_computed_fields)  # ty 0.0.83: Unknown
```

pyright and mypy both reveal `dict[str, FieldInfo]` / `dict[str, ComputedFieldInfo]`. The `invalid-type-form` ty reports is inside site-packages, so it is suppressed and the user only sees `Unknown`. Stripped of pydantic, it is the `after` shape from this thread:

```py
from __future__ import annotations

class C:
    def fields(self) -> dict[str, int]:
        return {}

    def dict(self) -> None: ...

reveal_type(C().fields())  # ty: Unknown; pyright, mypy, and `typing.get_type_hints` (3.12 and 3.14): dict[str, int]
```

(pydantic 2.13.5, ty 0.0.83)

**Comment by sharkdp** [comment]
> So the return annotation of `BaseModel.model_fields` and `model_computed_fields`, `dict[str, FieldInfo]`, resolves to that method

Hm, yes. That should really be fixed in Pydantic. Apparently they [are aware of that problem](https://github.com/pydantic/pydantic/blob/0384c970e37a59b344e75161eb106ea9996378ba/pydantic/main.py#L3-L5) but haven't applied that instruction consistently. I opened a PR upstream: https://github.com/pydantic/pydantic/pull/13853

**Comment by carljm** [comment]
This was recently discussed and resolved by the typing council in https://github.com/python/typing/pull/2144


I proposed there a compromise behavior that would be more compatible with the current behavior of pyright and mypy, but the rest of the typing council preferred to specify ty's current behavior, for simplicity and consistency. 

Now that this behavior is explicitly required by the typing spec, I don't think we will change it in ty. So closing this issue as "not planned".
